### ISSUE TITLE
Improve Device Initialisation Start Up Time

### ISSUE THREAD

**Issue by guerrillatiger** [fireice-uk/xmr-stak#1018]
Hi,

This issue is a duplicate of #201 however there was no resolution to it and it was closed.

Issue 
During initial startup xmr-stak is 'Compiling code and initialising GPUs', this takes about 5 seconds per thread. Is it possible to reduce this? When I want to change the coin I am mining I would like start up to be as quick as possible, this is approximately 1.5min.

Possible Solution 1:
Compile threads in parallel as my CPU only gets to make 30% during startup.

Possible Solution 2:
If no changes are made in the gpu configuration is there a need to compile again? Could this compiling be cached between starts, perhaps pass a value to say use cache if it exists.

Reproduction:
Start miner.

Rig Setup
OS: Windows 10 Pro 1709 16299.192
XMR Version: xmr-stak/2.2.0/c4400d19/master/win/nvidia-amd-cpu/aeon-monero/20
CPU: Intel Core i3-7100 (no CPU mining running)
GPU: AMD Vega 64 x8, undervolted and overclocked. Runs stable.

![xmr-stak slow start](https://user-images.githubusercontent.com/35990637/35612624-76646262-067a-11e8-8eeb-834dc8bd5f5c.PNG)

![xmr-stak cpu usage](https://user-images.githubusercontent.com/35990637/35612629-7ae2b9f6-067a-11e8-9501-41729f8fc5d4.PNG)

[amd.txt](https://github.com/fireice-uk/xmr-stak/files/1680905/amd.txt)

[clinfo_output.txt](https://github.com/fireice-uk/xmr-stak/files/1680907/clinfo_output.txt)

**Comment by silviotech** [comment]
I too am interested in this, i have a vega 64 FE, and created a little script to restart the program / disable and re-enable video card if necessary. Although i have quite stable settings currently sometimes it restarts numerous times. I thought it would be cool to have an option to pre- compile those settings for our specific AMD setup.

**Comment by khobbs19** [comment]
How about if config.txt has changed or keyboard input eg ‘x’ to re-read config.txt (not amd.txt etc) without restarting the miner? Thanks!

**Comment by eliashdezr** [comment]
Yep, hash the config files and if they don't change on each start, recompile for each GPU, this would decrease a lot the start time.

**Comment by jam3972** [comment]
For my 2x Vega FE rig, dev build from 2/19 of XMR-STAK takes 58 seconds from "Compiling Code" to "Starting AMD GPU thread" about 30 seconds per card.
It does this on every single run even without anything changing.

For comparison Cast-XMR 0.8.5 takes less than 1 second for this step on repeated runs. When it is a fresh driver installation it only takes about 10 seconds per card to compile.

So there is definitely room for optimization. Wish I knew how Cast-XMR did it.
Happy to provide any additional info/debug testing if it helps.

**Comment by psychocrypt** [comment]
XmrCast is only supporting rx5XX and vega and they are shipping precompiled
binaries for those gpus.The reason for the precompiled version is also that
we can not look to the opencl code.
Xmr-stak is supporting much more amd architectures and therefore we are
shipping the opencl code embedded in the binary. The  OpenCL code is
compiled explicitly for your device. This guarantee that xmr-stak can
support many devices(also non amd) and many different driver versions.

**Comment by psychocrypt** [comment]
What kind of cpu do you have. Xmr-stak compiles ~10 sec for my RX570 with a
AMD r5 1600X. If all gpus are equal than the binary is compiled only once
and used for both all gpus.

**Comment by jam3972** [comment]
Thanks for the explanation and all your great work. 
I am using an i3-4150 cpu. 
Both my devices are the same. AMD Vega FE cards. Unless I misunderstand, I don't think it is compiling only once. It does this every time without any changes to the cards:
output from build 7b850646:
```
[2018-02-22 12:12:11] : Compiling code and initializing GPUs. This will take a while...
[2018-02-22 12:12:11] : Device 0 work size 8 / 32.
[2018-02-22 12:12:29] : Device 0 work size 8 / 32.
[2018-02-22 12:12:50] : Device 1 work size 8 / 32.
[2018-02-22 12:13:06] : Device 1 work size 8 / 32.
[2018-02-22 12:13:25] : Starting AMD GPU thread 0, no affinity.
[2018-02-22 12:13:25] : Starting AMD GPU thread 1, no affinity.
[2018-02-22 12:13:25] : Starting AMD GPU thread 2, no affinity.
[2018-02-22 12:13:25] : Starting AMD GPU thread 3, no affinity.
```
here is my config:
```
"gpu_threads_conf" : [
  // gpu: gfx901 memory:15984
  // compute units: 64
  { "index" : 0,
    "intensity" : 2016, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : true, "mem_chunk" : 2,
    "comp_mode" : true
  },
  { "index" : 0,
    "intensity" : 2016, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : true, "mem_chunk" : 2,
    "comp_mode" : true
  },
  // gpu: gfx901 memory:15984
  // compute units: 64
  { "index" : 1,
    "intensity" : 2016, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : true, "mem_chunk" : 2,
    "comp_mode" : true
  },
  { "index" : 1,
    "intensity" : 2016, "worksize" : 8,
    "affine_to_cpu" : false, "strided_index" : true, "mem_chunk" : 2,
    "comp_mode" : true
  },
```

I can understand the long delay on first run, but the delays on repeated runs without any changes to the system I don't quite understand. Is there a way to save the compile results between runs?

**Comment by psychocrypt** [comment]
The miner compiles the OpenCL kernels each start once. Storing the compiled
kernel is not as easy. As soon as a new version of the miner is coming out
we need to handle that the kernel is recompiled. Also if the user changed
the system we need also to recompile. The logic to do this is not easy and
error prone.
Normally you start the miner only once and can run it many days. I know
that this is not possible with vega gpus but this is an driver issue. It is
an job for the vendor.

**Comment by jam3972** [comment]
OK Fair enough on saving the compile for repeated runs. I know the VEGA FEs have terrible drivers. 
let's focus on the one compile per run part first. You mention on the RX570 it compiles once and uses it for all cards.
For the VEGA FE is seems to be compiling once per card on every run. (it takes twice as long to start with 2 cards, and 6 times as long on a 6 card system) 
I'm not sure, but this isn't supposed to happen right? They are all the same card with the same config in amd.txt

**Comment by psychocrypt** [comment]
@guerrillatiger You can test the new pull request #1118. The PR compiles the code for each device in parallel.
I tested the code with a multi GPU NVIDIA system. The advantage of NVIDIA is that they use a compiler cash and the second start takes only 1 sec. If I empty the compiler cash it takes also ~20sec but due to the PR the time should be constant independent of the number of devices (if the CPU is fast).

**Comment by psychocrypt** [comment]
I implemented a compiler cache in #1131. This should help to speedup the miner start after the first restart.
The first start is not faster than the current miner version. The code must be copiled for each device and stored on disk. The second start should be fast.

**Comment by guerrillatiger** [comment]
@phychocrypt thanks. Sorry for the late reply but great solution, my 8 card Vega 64 rig is now about to start mining in 8 seconds.
